### ISSUE TITLE
Path priority

### ISSUE THREAD

**Issue by Ngob** [Kong/kubernetes-ingress-controller#203]
NOTE: GitHub issues are reserved for **bug reports only**.
For anything else, please join the conversation
in Kong Nation https://discuss.konghq.com/c/kubernetes.

---

### Summary

The matching priority between regex path and plain path looks broke. 

I am not sure if this is fully bug or a misunderstanding from me. From what I understood, If I have multiple ingress matching the same route, prefix path will have the highest priority. So if I have followings path: `/(.*)` and `/foo/baz/nurf`, the ingress for `/foo/baz/nurf` will match for the route "http://my-host.com/foo/baz/nurf". It looks like this is not always the case.

**Kong Ingress controller version**
0.2.2

**Kong or Kong Enterprise version**
0.14 CE

### Kubernetes version

**paste `kubectl version` output**
```
Client Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.7", GitCommit:"0c38c362511b20a098d7cd855f1314dad92c2780", GitTreeState:"clean", BuildDate:"2018-08-20T10:09:03Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"10+", GitVersion:"v1.10.6-gke.11", GitCommit:"42df8ec7aef509caba40b6178616dcffca9d7355", GitTreeState:"clean", BuildDate:"2018-11-08T20:06:00Z", GoVersion:"go1.9.3b4", Compiler:"gc", Platform:"linux/amd64"}
```
### Environment

- **Cloud provider or hardware configuration**: GKE
- **OS** (e.g. from /etc/os-release): COS
- **Kernel** (e.g. `uname -a`): Linux gke-prod-1-pool-8cpu-0cb67ce0-3x2f 4.14.56+ #1 SMP Sat Aug 11 23:13:59 PDT 2018 x86_64 Intel(R) Xeon(R) CPU @ 2.00GHz GenuineIntel GNU/Linux
- **Install tools**:
- **Others**:

### What happened

I am adding multiple ingress on the same host. Some use regex, some not. From the documentation, I understand [`Prefix paths are always evaluated first.`](https://docs.konghq.com/0.14.x/proxy/#matching-priorities), but it looks like this is not the case. I am adding following ingress (I added only relevant ingress):
```
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: my-main-ing
  namespace:  my-ns
  annotations:
    kubernetes.io/ingress.class: "nginx"
    plugins.konghq.com: my-ip-restriction
spec:
  tls:
  - hosts:
    - my-host.com
    secretName: my-tls-secret
  rules:
    - host: my-host.com
      http:
        paths:
          - path: "/(.*)"
            backend:
              serviceName: my-svc
              servicePort: https
---
apiVersion: configuration.konghq.com/v1
kind: KongIngress
metadata:
  name: my-main-ing
  namespace:  my-ns
route:
  strip_path: false
  preserve_host: true
  regex_priority: 0
  connect_timeout: 1000000
proxy:
  protocol: https
```
I do not know if that help, but as you can see I am using the IP restriction plugin on this ingress. After this, to be able to use certmanager, I added a dummy ingress. In my case, this is probably needed since I'm using the IP restriction plugin:
```
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: letsencrypt
  namespace:  my-ns
  annotations:
    kubernetes.io/ingress.class: "nginx"
spec:
  rules:
    - host: my-host.com
      http:
        paths:
          - path: "/.well-known"
            backend:
              serviceName: my-svc
              servicePort: https
---
apiVersion: configuration.konghq.com/v1
kind: KongIngress
metadata:
  name: letsencrypt
  namespace:  my-ns
route:
  strip_path: false
  preserve_host: true
  priority: 1000
```

For information, `my-svc` does exist. After this, I added my cert-manager certificate CRD. When I run `kubectl get ing letsencrypt -n my-ns -o yaml`, I get something like (I removed irrelevant parts):

```
  rules:
  - host: my-host.com
    http:
      paths:
      - backend:
          serviceName: my-svc
          servicePort: https
        path: /.well-known
      - backend:
          serviceName: cm-acme-http-solver-hash1
          servicePort: letsencrypt
        path: /.well-known/acme-challenge/hash2
```
Now when I run `curl "http://my-host.com/.well-known/acme-challenge/hash2"`, it looks like kong match `my-main-ing`. 
For information, if I run `curl "http://my-host.com/.well-known"` is proxied to the correct service

### Expected behavior

I would expect `curl "http://my-host.com/.well-known/acme-challenge/hash2"` to match the correct service (or at least the correct ingress). 

### Steps To Reproduce
I am using cert-manager so:
1. Install kubernetes + kong ingress
2. Install cert-manager
3. Deploy the main app
4. Create a `dummy` Ingress and a kong-ingress associated
5. Create the certificate using the dummy Ingress as the ingress use.

### Others

My current fix is to apply the `ip-restriction` plugin at a `Service` level (allowing me to remove the headache to have a dummy Ingress).

**Comment by hbagdi** [comment]
@Ngob Can you post the output of the following Admin API calls from Kong?
`GET /services`
`GET /routes`

We would like to inspect the values present in the DB to narrow down on what could be going wrong here.

**Comment by hbagdi** [comment]
@Ngob Friendly ping on this!

**Comment by Ngob** [comment]
Hello,

Sorry for the late reply, I have to reproduce the state on my preproduction environment, I will surely do it before the end of the week!
Thanks for your work!

**Comment by Tediferous** [comment]
Hello, @Ngob @hbagdi 

Any update on this issue? We are also experiencing path routing issues in our cluster. Intermittently kong will not route traffic correctly. Here is what our ingress looks like:

```
- apiVersion: extensions/v1beta1
  kind: Ingress
  metadata:
    annotations:
      kubernetes.io/ingress.class: nginx
  spec:
    rules:
    - host: abc.com
      http:
        paths:
        - backend:
            serviceName:  api
            servicePort: 80
          path: /api
        - backend:
            serviceName: web
            servicePort: 80
          path: /
    - host: def.com
      http:
        paths:
        - backend:
            serviceName: api
            servicePort: 80
          path: /api
        - backend:
            serviceName: web
            servicePort: 80
          path: /
    - host: ghi.com
      http:
        paths:
        - backend:
            serviceName: api
            servicePort: 80
          path: /api
        - backend:
            serviceName: web
            servicePort: 80
          path: /
```
We need all traffic  with a /api path to be routed to another service but this is not always the case. This setup works for some apps and not others

**Comment by hbagdi** [comment]
@Tediferous That doesn't seem to be the same issue as posted by OP.

What behavior do you see?
Do you see anything in the ingress-controller logs?

**Comment by Ngob** [comment]
Hello,

Sorry for the very late reply. I updated @hbagdi on slack about my API calls. 

For information, this previous configuration was to associate a kong plugin to a service or a ingress (to a route), and to associate another kong plugin to a service or a ingress (a sub-route). Now everything is working fine (until now) I do use one JWT plugin and working with anonymous consumer to restrict access via the IP plugin.

**Comment by hbagdi** [comment]
@Ngob Have you tested with Ingress Controller 0.3.0 and Kong 1.0.x?

**Comment by hbagdi** [comment]
@Tediferous Please open a separate issue to avoid hijacking the OP's problem

**Comment by crezy8** [comment]
@Tediferous 
I got the same problem, my ingress looks like yours.
kong: 1.0.0
kong ingress controller: 0.3.0
```
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: abc
  annotations:
    configuration.konghq.com: strip-path
spec:
  rules:
  - host: abc.com
    http:
      paths:
      - path: /api
        backend:
          serviceName: my-api-service
          servicePort: 80
      - path: /
        backend:
          serviceName: my-web-service
          servicePort: 80
```
The problem shows when I restart my api service, some uri like /api/xxx  can't pass  to my  api service and the others can. First, I found a way to solve problems by updating the ingress configuration(change anything -> update -> restore ingress config -> update again ). Finally I used a separate ingress to solve this problem, the ingress like these:
```
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: abc
  annotations:
    configuration.konghq.com: strip-path
spec:
  rules:
  - host: abc.com
    http:
      paths: 
      - path: /api
        backend:
          serviceName: my-api-service
          servicePort: 80
---
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: abc-frontend
  annotations:
    configuration.konghq.com: preserve-host
spec:
  rules:
  - host: abc.com
    http:
      paths: 
      - path: /
        backend:
          serviceName: my-web-service
          servicePort: 80
```
Maybe setting the Regex priority property will also work.

**Comment by hbagdi** [comment]
Fixed with Kong Ingress Controller 0.4.0.
